src/storage-client/src/sink.rs

```diff
@@ -474,25 +497,34 @@ pub async fn determine_latest_progress_record(
             return Ok(None);
         }
 
-        let mut latest_ts = None;
-        let mut latest_offset = None;
-
-        let progress_key_bytes = progress_key.as_bytes();
+        let mut maybe_upper = None;
+        let mut offset_upper = 0;
         while let Some((key, message, offset)) = get_next_message(progress_client, timeout)? {
-            debug_assert!(offset >= latest_offset.unwrap_or(0));
-            latest_offset = Some(offset);
-
-            let timestamp_opt = if &key == progress_key_bytes {
-                let progress: ProgressRecord = serde_json::from_slice(&message)?;
-                Some(progress.timestamp)
-            } else {
-                None
+            assert!(offset_upper <= offset);
+            offset_upper = offset + 1;
+
+            if &key != progress_key.as_bytes() {
+                continue;
+            }
+
+            let new_upper = match serde_json::from_slice::<ProgressRecord>(&message) {
```

Comment:
this will never fail, it will produce an empty frontier if presented with a record that is _missing_ a `frontier` field, and ignore the `timestamp` field

I could have sworn this was configurable with `serde_json` but i may be wrong; probably something like https://stackoverflow.com/questions/76645782/rust-serde-deserialization-how-can-i-make-sure-a-field-does-not-exist-and-throw _with unit tests_ is the right route here. im surprised upgrade tests didnt catch this...

Reply:
I don't think that's the case https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4a77bf00ebe417dc7da349f1818cf39c

You have to put annotations if you want serde to fill in fields that are missing. Maybe you're thinking of optional fields?

Reply:
I added a unit test nevertheless

Reply:
ah i see, it doesnt default to an empty vec! thanks for the unit test!